Async paste2

tastybento opened the pull request:
This is a placeholder because it needs more work. The "happy path" works, but it's not yet optimized and the edge case of the player logging out mid-paste needs investigation.

In summary, a slightly different schem format is required to optimize the pasting.

tastybento committed:
WIP Added staggered pasting

tastybento committed:
Merge branch 'develop' into async-paste2

tastybento committed:
WIP pastes but attachments fall when split across paste boundaries

tastybento committed:
WIP: Enables async-style pasting of schems.

Speed of schem pasting is in the config.yml of BentoBox.

Old schems will work and paste async, but attached blocks, e.g.,
torches, may fall off due to being pasted ticks before the supporting
block is pasted. Newer schems will not have this issue.

Further work is needed to optimize entity pasting.

Known issue: if a player logs out during the pasting, things go wrong
until the next reload. Needs investigation and mitigation.

There's no mitigation against the server shutting down or crashing
mid-paste.

tastybento commented on src/main/java/world/bentobox/bentobox/BentoBox.java:
```diff
@@ -201,10 +201,7 @@ public void onDisable() {
-        if (settings != null) {
-            new Config<>(this, Settings.class).saveConfigObject(settings);
-        }
+        // Save settings - not required
```
This has nothing to do with the paste - I just realized it's not needed (yet)

Poslovitch replied:
What was it expected to do?

tastybento replied:
I don’t know. I think it was left over from when BSB was part of the plugin. In theory, if any of the settings were changed in-game then we’d need it.

tastybento replied:
This needs to go back in like I said. Oops. It's there to save new settings to config.

tastybento commented on src/main/java/world/bentobox/bentobox/managers/IslandsManager.java:
```diff
@@ -679,25 +680,23 @@ public void removePlayer(World world, UUID uuid) {
     public void removePlayersFromIsland(Island island) {
-        // Teleport players away
-        for (Player player : Bukkit.getOnlinePlayers()) {
-            if (island.inIslandSpace(player.getLocation().getBlockX(), player.getLocation().getBlockZ())) {
```
This is actually a bug. It does not consider the world, so it could teleport players in other worlds! Oops.

Poslovitch replied:
Good catch!

tastybento commented on src/main/java/world/bentobox/bentobox/managers/island/NewIsland.java:
```diff
@@ -165,6 +162,12 @@ public void newIsland() {
+            if (oldIsland != null) {
+                // Delete the old island
+                plugin.getIslands().deleteIsland(oldIsland, true);
+            }
```
Another bug: The ordering was wrong before. Island was being deleted before the player's island was pasted so they ended up in the acid sea.

tastybento commented on src/main/java/world/bentobox/bentobox/schems/Clipboard.java:
```diff
@@ -55,8 +58,11 @@
 public class Clipboard {
 
+    // Speed of pasting
+    private int pasteSpeed = 200;
```
This default should never be used actually.

tastybento commented on src/main/java/world/bentobox/bentobox/schems/Clipboard.java:
```diff
@@ -188,14 +197,44 @@ public void pasteIsland(World world, Island island, Runnable task) {
-            blockConfig.getConfigurationSection(BLOCK).getKeys(false).forEach(b -> pasteBlock(world, island, loc, blockConfig.getConfigurationSection(BLOCK + "." + b)));
+            Iterator<String> it = blockConfig.getConfigurationSection(BLOCK).getKeys(false).iterator();
+            Iterator<String> attachmentsIt = blockConfig.contains(ATTACHED) ? blockConfig.getConfigurationSection(ATTACHED).getKeys(false).iterator() : null;
+            pastingTask = Bukkit.getScheduler().runTaskTimer(plugin, () -> {
```
I intend to add another Iterator for Entities. Right now they are included in the blocks, but should have their own YAML section in the schem file.

tastybento commented on src/main/java/world/bentobox/bentobox/schems/Clipboard.java:
```diff
@@ -257,14 +296,10 @@ private void pasteBlock(World world, Island island, Location location, Configura
         Block block = world.getBlockAt(x, y, z);
         String blockData = config.getString("bd");
         if (blockData != null) {
-            if (config.getBoolean(ATTACHED)) {
```
This whole section was not actually ever used and was from an earlier version of schems. So I deleted it.

tastybento commented on src/main/java/world/bentobox/bentobox/schems/Clipboard.java:
```diff
@@ -257,14 +296,10 @@ private void pasteBlock(World world, Island island, Location location, Configura
-        // Entities
-        if (config.isConfigurationSection(ENTITY)) {
+        // Entities - only paste here if this is a clipboard paste, otherwise, wait until island is pasted
+        if (pasteEntities && config.isConfigurationSection(ENTITY)) {
```
Pasting using the clipboard is NOT done async right now, just when pasting an island in the game. So using the schem paste command may lag the server. This will need changing in the future once the async pasting works completely.

tastybento replied:
Now it's done async too.

tastybento commented on src/main/java/world/bentobox/bentobox/schems/Clipboard.java:
```diff
@@ -285,7 +320,7 @@ private void setEntity(Location location, ConfigurationSection config) {
         en.getKeys(false).forEach(k -> {
             ConfigurationSection ent = en.getConfigurationSection(k);
-            Location center = location.add(new Vector(0.5, 0.0, 0.5));
+            Location center = location.add(new Vector(0.5, 1.0, 0.5));
```
Bug: Pasting entities inside the block they were found, is not the right way.

tastybento replied:
Adjusted a bit.

tastybento commented on src/main/java/world/bentobox/bentobox/schems/Clipboard.java:
```diff
@@ -413,15 +448,35 @@ private boolean copyBlock(Block block, Location copyOrigin, boolean copyAir, Col
+
         // Set block data
-        s.set("bd", block.getBlockData().getAsString());
+        if (bs.getData() instanceof Attachable) {
```
getData() returns MaterialData, which is deprecated, but there is no other way to determine Attachable, so it has to be used.

tastybento commented on src/main/java/world/bentobox/bentobox/schems/Clipboard.java:
```diff
@@ -413,15 +448,35 @@ private boolean copyBlock(Block block, Location copyOrigin, boolean copyAir, Col
         // Set block data
-        s.set("bd", block.getBlockData().getAsString());
+        if (bs.getData() instanceof Attachable) {
+            plugin.logDebug("Attachable");
```
Debug needs to be removed.

tastybento commented on src/main/java/world/bentobox/bentobox/schems/Clipboard.java:
```diff
@@ -413,15 +448,35 @@ private boolean copyBlock(Block block, Location copyOrigin, boolean copyAir, Col
+            ConfigurationSection a = blockConfig.createSection(ATTACHED + pos);
+            a.set("bd", block.getBlockData().getAsString());
+            // Placeholder
+            s.set("bd", "minecraft:stone");
```
Attachment blocks need to be set to something that doesn't have to be attached, e.g. STONE. If you do not do this and there is water or lava there, then that liquid will flow between ticks and mess everything up. It could be changed to AIR I think maybe and still be the same.

tastybento commented on src/main/resources/config.yml:
```diff
@@ -75,6 +75,9 @@ island:
     max-length: 20
+  # Number of blocks to paste per tick when pasting a schem
+  # Smaller values will help reduce noticeable lag but will make pasting take longer
+  paste-speed: 1000
```
I have no idea if this is a good speed or not.

Poslovitch replied:
Hopefully we'll get some feedback about that.

tastybento commented on src/main/resources/locales/en-US.yml:
```diff
@@ -246,7 +246,7 @@ commands:
       too-many-islands: "&cThere are too many islands in this world: there isn't enough room for yours to be created."
       unable-create-island: "&cYour island could not be generated, please contact an administrator."
-      creating-island: "&aCreating your island..."
+      creating-island: "&aCreating your island, please wait a moment..."
```
Now that it is async, it may take some time.

tastybento committed:
Rewrote using state approach to be easier to understand.

Backwards compatible with current schems. New schems required to avoid
torch popping etc. Look for updates to AcidIsland and BSB.

tastybento committed:
Removed unneeded enum values. Improved const names.

tastybento committed:
Merge remote-tracking branch 'origin/develop' into async-paste2

tastybento committed:
Final touches

tastybento committed:
Putting the config saving back in.

tastybento committed:
Fixes tests.

I made pasting the clipboard async as well so this affected the test.

tastybento committed:
Added null checking protection.

This may help if a database is becoming corrupted.

tastybento committed:
Merge remote-tracking branch 'origin/develop' into async-paste2

tastybento committed:
Fixes double home teleport when resetting island.
